File: basics/src/main/java/javaprograms/sorting/Sort.java
package javaprograms.sorting;

import java.util.Comparator;

/**
 * This interface is useful in defining the instances of [[java.util.Comparator]]
 * for custom classes.
 */
interface Sort {
    Comparator<Double> doubleComparator = (o1, o2) -> (int)((o1 - o2) * -1);
}

File: basics/src/main/scala/scalaprograms/sorting/Sort.scala
package scalaprograms.sorting

/**
  * This trait is helpful in defining any implicits required for sorting.
  * e.g. the [[Ordering]] object for custom classes, whenever used.
  *
  * The algorithms are implemented with Functional Programming paradigm in mind.
  */
trait Sort {
  def sort[T](in: Seq[T])(implicit ord: Ordering[T]): Seq[T]
}
